dd-java-agent/instrumentation/aws-java-sns-2.0/src/main/java/datadog/trace/instrumentation/aws/v2/sns/SnsInterceptor.java

```diff
@@ -48,22 +59,38 @@ public SdkRequest modifyRequest(
       // 10 messageAttributes is a limit from SQS, which is often used as a subscriber, therefore
       // the limit still applies here
       if (request.messageAttributes().size() < 10) {
+        // Get topic name for DSM
+        Optional<String> snsTopicArn = request.getValueForField("TopicArn", String.class);
+        if (!snsTopicArn.isPresent()) {
+          snsTopicArn = request.getValueForField("TargetArn", String.class);
+        }
+        Optional<String> snsTopicName =
+            snsTopicArn.map(arn -> arn.substring(arn.lastIndexOf(':') + 1));
```

Comment:
why all the fuss with Optional and not some straightforward code like you did for v1 ?
```suggestion
        String snsTopicName = request.topicArn() == null ? request.targetArn() : request.topicArn();
```

Reply:
The targetARN isn't available as a field for `PublishBatchRequest` but I think we can skip the check since topicName should be valid. Made this change.